### PR TITLE
[13.0][MIG] connector_ecommerce

### PR DESCRIPTION
**Pull request by bizzappdev**
Based on https://github.com/OCA/connector-ecommerce/pull/53/files with new fixes

### CODE REVIEW

**Comment by StefanRijnhart** [comment]
Thanks! Can you have a look at the broken onchange tests?
```
2021-03-04 10:56:59,684 23235 INFO openerp_test odoo.addons.connector_ecommerce.tests.test_onchange: Starting TestOnchange.test_play_onchange ...
2021-03-04 10:57:00,424 23235 INFO openerp_test odoo.addons.connector_ecommerce.tests.test_onchange: ======================================================================
2021-03-04 10:57:00,424 23235 ERROR openerp_test odoo.addons.connector_ecommerce.tests.test_onchange: ERROR: TestOnchange.test_play_onchange
Traceback (most recent call last):
File "/home/travis/build/OCA/connector-ecommerce/connector_ecommerce/tests/test_onchange.py", line 119, in test_play_onchange
self.assertEqual(line["tax_id"], [(5,), (4, tax.id)])
KeyError: 'tax_id'
```

**Comment by bizzappdev** [comment]
@StefanRijnhart Tax_id test case has been resolved.

**Comment by PCatinean** [comment]
@gurneyalex I tried rebuilding the runbot build but it still has a warning which fails the PR and I can't view the logs from the build to find out what is the issue. Could you help out?

**Review by PCatinean — APPROVED** [review]

**Comment by PCatinean** [comment]
Tests are passing can you please review @StefanRijnhart ?

**Review by sbejaoui — APPROVED** [review]

**Comment by bizzappdev** [comment]
@StefanRijnhart all seems okay. can we proceed with a merge? or still some suggested changes?

File: connector_ecommerce/tests/test_onchange.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    def test_play_onchange(self):
+        """ Play the onchange ConnectorUnit on a sales order """
+        product_model = self.env["product.product"]
+        partner_model = self.env["res.partner"]
+        tax_model = self.env["account.tax"]
+
+        partner = partner_model.create(
+            {"name": "seb", "zip": "69100", "city": "Villeurbanne"}
+        )
```
Comment (victoralmau) [inline connector_ecommerce/tests/test_onchange.py]:
```suggestion
        fiscal_position = self.env["account.fiscal.position"].create({
            "name": "Test", "company_id": self.env.ref("base.main_company").id
        })
        
        partner = partner_model.create({
            "name": "seb",
            "zip": "69100",
            "city": "Villeurbanne",
            "property_account_position_id": fiscal_position.id
        })
```

File: connector_ecommerce/tests/test_onchange.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                    "sequence": 2,
+                },
+            ),
+        ]
+
+        with self.get_base() as base:
+            onchange = base.component(usage="ecommerce.onchange.manager.sale.order")
+            order = onchange.play(order_vals, extra_lines)
+
+        self.assertEqual(order["partner_invoice_id"], partner_invoice.id)
```
Comment (victoralmau) [inline connector_ecommerce/tests/test_onchange.py:125]:
```suggestion
        self.assertEqual(order["fiscal_position_id"], fiscal_position.id)
        self.assertEqual(order["partner_invoice_id"], partner_invoice.id)
```

File: connector_ecommerce/components/line_builder.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        builder.get_line()
+
+    """
+
+    _name = "ecommerce.order.line.builder"
+    _inherit = "base.connector"
+    _usage = "order.line.builder"
+
+    def __init__(self, work_context):
+        super(SpecialOrderLineBuilder, self).__init__(work_context)
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        super().__init__(work_context)
```

File: connector_ecommerce/components/line_builder.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+class ShippingLineBuilder(Component):
+    """ Return values for a Shipping line """
+
+    _name = "ecommerce.order.line.builder.shipping"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.shipping"
+
+    def __init__(self, work_context):
+        super(ShippingLineBuilder, self).__init__(work_context)
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        super().__init__(work_context)
```

File: connector_ecommerce/components/line_builder.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        self.product_ref = ("connector_ecommerce", "product_product_shipping")
+        self.sequence = 999
+
+    def get_line(self):
+        values = super(ShippingLineBuilder, self).get_line()
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        values = super().get_line()
```

File: connector_ecommerce/components/line_builder.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+class CashOnDeliveryLineBuilder(Component):
+    """ Return values for a Cash on Delivery line """
+
+    _name = "ecommerce.order.line.builder.cod"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.cod"
+
+    def __init__(self, work_context):
+        super(CashOnDeliveryLineBuilder, self).__init__(work_context)
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        super().__init__(work_context)
```

File: connector_ecommerce/components/line_builder.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+class GiftOrderLineBuilder(Component):
+    """ Return values for a Gift line """
+
+    _name = "ecommerce.order.line.builder.gift"
+    _inherit = "ecommerce.order.line.builder"
+    _usage = "order.line.builder.gift"
+
+    def __init__(self, work_context):
+        super(GiftOrderLineBuilder, self).__init__(work_context)
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        super().__init__(work_context)
```

File: connector_ecommerce/components/line_builder.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        self.product_ref = ("connector_ecommerce", "product_product_gift")
+        self.sign = -1
+        self.gift_code = None
+        self.sequence = 990
+
+    def get_line(self):
+        line = super(GiftOrderLineBuilder, self).get_line()
```
Comment (victoralmau) [inline connector_ecommerce/components/line_builder.py]:
```suggestion
        line = super().get_line()
```

File: connector_ecommerce/models/product.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            # to fire the event 2 times
+            if self.env.context.get("from_product_ids"):
+                from_product_ids = self.env.context["from_product_ids"]
+                remove_products = product_model.browse(from_product_ids)
+                products -= remove_products
+            for product in products:
+                self._event("on_product_price_changed").notify(product)
+
+    def write(self, vals):
+        result = super(ProductTemplate, self).write(vals)
```
Comment (victoralmau) [inline connector_ecommerce/models/product.py]:
```suggestion
        result = super().write(vals)
```

File: connector_ecommerce/models/product.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+    def write(self, vals):
+        self_context = self.with_context(from_product_ids=self.ids)
+        result = super(ProductProduct, self_context).write(vals)
+        self._price_changed(vals)
+        return result
+
+    @api.model
+    def create(self, vals):
+        product = super(ProductProduct, self).create(vals)
```
Comment (victoralmau) [inline connector_ecommerce/models/product.py]:
```suggestion
        product = super().create(vals)
```

File: connector_ecommerce/models/sale.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+                % order.name
+            )
+            for picking in order.picking_ids:
+                picking.message_post(body=message)
+            for invoice in order.invoice_ids:
+                invoice.message_post(body=message)
+
+    @api.model
+    def create(self, values):
+        order = super(SaleOrder, self).create(values)
```
Comment (victoralmau) [inline connector_ecommerce/models/sale.py]:
```suggestion
        order = super().create(values)
```

File: connector_ecommerce/models/sale.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        if values.get("canceled_in_backend"):
+            order._log_canceled_in_backend()
+            order._try_auto_cancel()
+        return order
+
+    def write(self, values):
+        result = super(SaleOrder, self).write(values)
```
Comment (victoralmau) [inline connector_ecommerce/models/sale.py]:
```suggestion
        result = super().write(values)
```

File: connector_ecommerce/models/sale.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        if values.get("canceled_in_backend"):
+            self._log_canceled_in_backend()
+            self._try_auto_cancel()
+        return result
+
+    def action_cancel(self):
+        res = super(SaleOrder, self).action_cancel()
```
Comment (victoralmau) [inline connector_ecommerce/models/sale.py]:
```suggestion
        res = super().action_cancel()
```

File: connector_ecommerce/models/sale.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        action["res_id"] = parent.id
+        return action
+
+    def _create_delivery_line(self, carrier, price_unit):
+        if self.order_line.filtered(lambda r: r.is_delivery):
+            # skip if we have already a delivery line (created by
+            # import of order)
+            return
+        else:
+            return super(SaleOrder, self)._create_delivery_line(carrier, price_unit)
```
Comment (victoralmau) [inline connector_ecommerce/models/sale.py]:
```suggestion
            return super()._create_delivery_line(carrier, price_unit)
```

File: connector_ecommerce/models/stock.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    _inherit = "stock.picking"
+
+    related_backorder_ids = fields.One2many(
+        comodel_name="stock.picking",
+        inverse_name="backorder_id",
+        string="Related backorders",
+    )
+
+    def write(self, vals):
+        res = super(StockPicking, self).write(vals)
```
Comment (victoralmau) [inline connector_ecommerce/models/stock.py]:
```suggestion
        res = super().write(vals)
```

File: connector_ecommerce/models/stock.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def action_done(self):
+        fire_event = not self.env.context.get("__no_on_event_out_done")
+        if fire_event:
+            pickings = self.mapped("picking_id")
+            states = {p.id: p.state for p in pickings}
+
+        result = super(StockMove, self).action_done()
```
Comment (victoralmau) [inline connector_ecommerce/models/stock.py]:
```suggestion
        result = super().action_done()
```

File: connector_ecommerce/tests/test_invoice_event.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+import mock
+
+import odoo.tests.common as common
+
+
+class TestInvoiceEvent(common.TransactionCase):
+    """ Test if the events on the invoice are fired correctly """
+
+    def setUp(self):
+        super(TestInvoiceEvent, self).setUp()
```
Comment (victoralmau) [inline connector_ecommerce/tests/test_invoice_event.py]:
```suggestion
        super().setUp()
```

File: connector_ecommerce/tests/test_onchange.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+DB = common.DB
+ADMIN_USER_ID = common.ADMIN_USER_ID
+
+
+class TestOnchange(TransactionComponentRegistryCase):
+    """ Test if the onchanges are applied correctly on a sales order"""
+
+    def setUp(self):
+        super(TestOnchange, self).setUp()
```
Comment (victoralmau) [inline connector_ecommerce/tests/test_onchange.py]:
```suggestion
        super().setUp()
```

File: connector_ecommerce/tests/test_picking_event.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        vals.update(**values)
+        pack_operation = move_line_env.new(vals)
+        pack_operation.onchange_product_id()
+        return move_line_env.create(
+            pack_operation._convert_to_write(pack_operation._cache)
+        )
+
+    def setUp(self):
+        super(TestPickingEvent, self).setUp()
```
Comment (victoralmau) [inline connector_ecommerce/tests/test_picking_event.py]:
```suggestion
        super().setUp()
```

**Review by victoralmau — CHANGES_REQUESTED** [review]
Good job.
Some little changes and add some things to sale order test (create fiscal position to set in partner and check if order set these fiscal position).

**Comment by bizzappdev** [comment]
Sorry for delay. Added suggested changes.

**Comment by gurneyalex** [comment]
hello @bizzappdev it seems the module was merged on 13.0. You mention some additional fixes, could you make another PR to include these fixes on top of the current 13.0 branch?

**Comment by bizzappdev** [comment]
> hello @bizzappdev it seems the module was merged on 13.0. You mention some additional fixes, could you make another PR to include these fixes on top of the current 13.0 branch?

@gurneyalex Fixes are also taken care in #62 so we can close this PR

**Pull request closed without merging**